Refactor retry logic for TransportMasterNodeAction

ywelsch opened the pull request:
Changes:
- Introduce inner class `AsyncSingleAction` to store state used across retry invocations.
- Same as for `TransportInstanceSingleOperationAction` and `TransportReplicationAction`, `onClusterServiceClose` consistently throws a `NodeClosedException` now.
- Added retry logic if master could not publish cluster state or stepped down before publishing (ZenDiscovery). The test `IndexingMasterFailoverIT` shows the issue (but will not be part of the final commit).
- Simplified retry logic by moving bits from different places into shared retry method.
  - Removed boolean flag `retrying` that aborted retrying after a single master node change (now we retry until timeout).
  - Two existing predicates that deal with master node changes unified in a single predicate `masterNodeChangedPredicate`.

Other implementation notes: In contrast to `TransportInstanceSingleOperationAction` and `TransportReplicationAction`, `onTimeout` does not try to execute the action a last time as in this case we wait for specific cluster change events. This keeps the behaviour consistent with the old code that directly failed on timeout.

ywelsch commented:
The method `processBeforeDelegationToMaster` can also be removed. AFAICS, it is nowhere used.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java:
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
-            transportService.sendRequest(nodes.masterNode(), actionName, request, new BaseTransportResponseHandler<Response>() {
+        }
+
+        private ClusterStateObserver.ChangePredicate retryableOrNoBlockPredicate() {
```
can we make these final fields on the parent class? they don't have any state.

ywelsch replied:
retryableOrNoBlockPredicate depends on the current request being made. I chose a method so that the predicate object does not have to be created for every MasterNodeAction.

bleskes replied:
I see. I don't think it's that bad to create it for every (master) request and keep it simple but up to you . If you leave like this please add a comment in the code as to why. Also, my comment referred to both this and masterNodeChangedPredicate , so fi you decided to change, please do it there too.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java:
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
+                        @Override
+                        public void onFailure(Throwable t) {
+                            if (t instanceof Discovery.FailedToCommitClusterStateException
+                                    || (t instanceof EsRejectedExecutionException && t.getMessage().startsWith("no longer master."))) {
```
lets' introduce a dedicated exception for this. We can upgrade discovery.zen.NotMasterException to be in the cluster package and use that.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java:
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
+                        public void onFailure(Throwable t) {
+                            if (t instanceof Discovery.FailedToCommitClusterStateException
+                                    || (t instanceof EsRejectedExecutionException && t.getMessage().startsWith("no longer master."))) {
+                                logger.debug("master could not publish cluster state or stepped down before publishing, scheduling a retry");
```
can we add the action to the log so we know which operation this is about?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java:
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
-                                }
+                    retry(new MasterNotDiscoveredException(), masterNodeChangedPredicate());
+                } else {
+                    processBeforeDelegationToMaster(request, clusterState);
```
do we still need this method?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java:
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
+                            Throwable cause = exp.unwrapCause();
+                            if (cause instanceof ConnectTransportException) {
+                                // we want to retry here a bit to see if a new master is elected
+                                logger.debug("connection exception while trying to forward request to master node [{}], scheduling a retry. Error: [{}]",
```
can we add the action name here too?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java:
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
+                    retry(new MasterNotDiscoveredException(), masterNodeChangedPredicate());
+                } else {
+                    processBeforeDelegationToMaster(request, clusterState);
+                    transportService.sendRequest(nodes.masterNode(), actionName, request, new BaseTransportResponseHandler<Response>() {
```
using ActionListenerResponseHandler will simplify this lightly.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java:
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
-                    listener.onResponse(response);
+                public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                                     ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+                    return newState.nodes().masterNodeId() != null && (previousState != newState || previousStatus != newStatus);
```
I know this is how it used to be, but can we make the if be more like the `masterNodeChangePredicate` name and check the the master node is not null and have changed? (we now test for a cluster state change)

bleskes commented on core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java:
```diff
@@ -81,146 +83,150 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
+
+                    @Override
+                    public void onTimeout(TimeValue timeout) {
+                        listener.onFailure(failure);
```
can we add a timeout log here?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java:
```diff
@@ -0,0 +1,140 @@
+        return classes;
+    }
+
+    public void testMasterFailoverDuringIndexingWithMappingChanges() throws Throwable {
```
why remove this test?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java:
```diff
@@ -0,0 +1,140 @@
+                .build();
+
+        for (int i = 0; i < 3; i++) {
+            internalCluster().startMasterOnlyNode(sharedSettings);
```
startNodesAsync(3,shardSettings)?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java:
```diff
@@ -0,0 +1,140 @@
+
+        logger.info("--> wait for all nodes to join the cluster");
+
+        Client client = client();
```
why cache the client?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java:
```diff
@@ -0,0 +1,140 @@
+        client(dataNode).prepareIndex("myindex", "mytype").setSource("field1", "val1").get();
+
+        final BlockingQueue<Throwable> exceptions = new LinkedBlockingQueue();
+        Thread[] threads = new Thread[6];
```
can't we make do with one background thread?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java:
```diff
@@ -0,0 +1,140 @@
+
+        for (int i = 0; i < 10; i++) {
+            // periodically interrupt communication between master and other nodes in cluster
+            Thread.sleep(1000);
```
boo hoe! Let's start a disruption, wait for a new master via the other nodes, stop the disruption and repeat. Why the sleep?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java:
```diff
@@ -0,0 +1,140 @@
+            threads[j].setName("t_" + j);
+            threads[j].start();
+        }
+
```
I think we need a barrier.await() here too to make sure the threads and disruptions start together?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java:
```diff
@@ -0,0 +1,140 @@
+                            assertThat(response.isCreated(), equalTo(true));
+                        }
+                    } catch (Throwable t) {
+                        exceptions.offer(t);
```
we can throw exceptions on the background thread and the test should fail with uncaught exceptions on a thread.

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java:
```diff
@@ -0,0 +1,140 @@
+
+        for (Thread t : threads) t.join();
+
+        if (!exceptions.isEmpty()) {
```
can we also verify all the documents are found in the index?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java:
```diff
@@ -320,4 +322,35 @@ protected void processBeforeDelegationToMaster(Request request, ClusterState sta
+
+        clusterService.setState(ClusterStateCreationUtils.state(localNode, localNode, allNodes));
+
+        new Action(Settings.EMPTY, "testAction", transportService, clusterService, threadPool) {
```
can we increase the timeout on the request? if one runs a debugger the test may fail to retry because a timeout happens, making it confusing.

bleskes replied:
ping?

ywelsch replied:
Maybe I misunderstood something here, but isn't that done by the first line in the method?
```
Request request = new Request().masterNodeTimeout(TimeValue.timeValueHours(1));
```

bleskes replied:
oops. Missed it. You're right.

bleskes commented:
This looks great. Left a bunch of comments.

ywelsch commented:
@bleskes, can you have another look?

I made (all) changes according to the provided suggestions, in particular:

- removed processBeforeDelegationToMaster
- changed methods retryableOrNoBlockPredicate() and masterNodeChangedPredicate() to fields
- changed masterNodeChangedPredicate implementation to only trigger on master node id changes
- moved discovery.zen.NotMasterException to cluster.NotMasterException
- log output in case of timeout while retrying (log-level warn)
- simplified IntegrationTest (took settings from DiscoveryWithServiceDisruptionsIT)

bleskes commented on core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java:
```diff
@@ -48,6 +50,19 @@
  * A base class for operations that needs to be performed on the master node.
  */
 public abstract class TransportMasterNodeAction<Request extends MasterNodeRequest, Response extends ActionResponse> extends HandledTransportAction<Request, Response> {
+    private static final ClusterStateObserver.ChangePredicate masterNodeChangedPredicate = new ClusterStateObserver.ChangePredicate() {
```
nit: can we move this next to the other predicate? I think it's good to have them together.

ywelsch replied:
yes, but then it cannot be static anymore (as inner class does not allow static member)

bleskes replied:
ok. fair enough..

> On 04 Nov 2015, at 13:22, Yannick Welsch <notifications@github.com> wrote:
> 
> In core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java:
> 
> > @@ -48,6 +50,19 @@
> >   * A base class for operations that needs to be performed on the master node.
> >   */
> >  public abstract class TransportMasterNodeAction<Request extends MasterNodeRequest, Response extends ActionResponse> extends HandledTransportAction<Request, Response> {
> > +    private static final ClusterStateObserver.ChangePredicate masterNodeChangedPredicate = new ClusterStateObserver.ChangePredicate() {
> 
> yes, but then it cannot be static anymore (as inner class does not allow static member)
> 
> —
> Reply to this email directly or view it on GitHub.
>

ywelsch replied:
fair enough, leave it or fair enough, change it ;-)

bleskes replied:
fair enough, leave it.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java:
```diff
@@ -75,152 +90,125 @@ protected boolean localExecute(Request request) {
-                        listener.onFailure(exp);
+                    @Override
+                    public void onTimeout(TimeValue timeout) {
+                        logger.warn("timed out while retrying. Reason for retry: ", failure);
```
Somewhat simpler: ("timed out while retrying [{}] after failure (timeout [{}])", action, failure) . I'm doubting between DEBUG and WARN for this log...

ywelsch replied:
I was also torn between DEBUG and WARN. If you prefer, I can make it DEBUG for now, and we can switch it to WARN if that is useful for operations.

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java:
```diff
@@ -0,0 +1,119 @@
+import static org.hamcrest.Matchers.equalTo;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+@ESIntegTestCase.SuppressLocalMode
```
I don't think we need this? we want to make sure we use zen, but that's it no?

ywelsch replied:
If I remove SuppressLocalMode, and add ("discovery.type", "zen") to the test, I hit the issue with zendiscovery + local mode again:

```
[2015-11-04 14:38:31,931][WARN ][org.elasticsearch.discovery] [node_t0] waited for 30s and no initial state was set by the discovery
[2015-11-04 14:38:31,932][INFO ][org.elasticsearch.node   ] [node_t0] started
[2015-11-04 14:38:32,042][WARN ][org.elasticsearch.discovery] [node_t1] waited for 30s and no initial state was set by the discovery
[2015-11-04 14:38:32,043][INFO ][org.elasticsearch.node   ] [node_t1] started
[2015-11-04 14:38:32,146][WARN ][org.elasticsearch.discovery] [node_t2] waited for 30s and no initial state was set by the discovery
```

and the following exception at the end:

```
MasterNotDiscoveredException[]
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.doStart(TransportMasterNodeAction.java:167)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.start(TransportMasterNodeAction.java:123)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:81)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:52)
	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:70)
	at org.elasticsearch.action.support.HandledTransportAction$TransportHandler.messageReceived(HandledTransportAction.java:47)
	at org.elasticsearch.action.support.HandledTransportAction$TransportHandler.messageReceived(HandledTransportAction.java:43)
	at org.elasticsearch.transport.local.LocalTransport.handleRequest(LocalTransport.java:290)
	at org.elasticsearch.transport.local.LocalTransport.messageReceived(LocalTransport.java:251)
	at org.elasticsearch.transport.local.LocalTransport$1.run(LocalTransport.java:229)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

I am still investigating that one...

bleskes replied:
argh, I forgot UnicastZenPing doesn't support local addresses by default. Sorry for the noise..  (we should fix this at one point)

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java:
```diff
@@ -0,0 +1,119 @@
+                .put(FaultDetection.SETTING_PING_RETRIES, "1") // for hitting simulated network failures quickly
+                .put("discovery.zen.join_timeout", "10s")  // still long to induce failures but to long so test won't time out
+                .put(DiscoverySettings.PUBLISH_TIMEOUT, "1s") // <-- for hitting simulated network failures quickly
+                .put("http.enabled", false) // just to make test quicker
```
I think we can clean up the http/transport/gatway settings here

ywelsch replied:
If I remove 

```
.put("transport.host", "127.0.0.1") // only bind on one IF we use v4 here by default
```

I run into #14513

bleskes replied:
sure. I hope #14513 will be fixed soon. Not sure if you that is what you meant, but I think we better wait rather then work around it.

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java:
```diff
@@ -0,0 +1,119 @@
+        ensureStableCluster(4);
+
+        // We index data with mapping changes into cluster and have master failover at same time
+        client(dataNode).prepareIndex("myindex", "mytype").setSource("field1", "val1").get();
```
I think we can simplify this and make sure we have 1 shard, no replicas.

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java:
```diff
@@ -0,0 +1,119 @@
+                try {
+                    barrier.await();
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
```
boo! :) just ignore :)

bleskes replied:
if you want to know i happened, let's use logging..

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java:
```diff
@@ -0,0 +1,119 @@
+                    e.printStackTrace();
+                    return;
+                }
+                for (int i = 0; i < 10; i++) {
```
is 10 iteration enough to reproduce reliably without your fix?

ywelsch replied:
It triggered FailedToCommitClusterStateException every time on my machine.

bleskes replied:
++

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterFailoverIT.java:
```diff
@@ -0,0 +1,119 @@
+        indexingThread.setName("indexingThread");
+        indexingThread.start();
+
+        ensureStableCluster(4);
```
why do this again?

ywelsch replied:
extra stability (lol no, a relic from an earlier iteration)

bleskes commented:
Update looks great. I left some minor comments mostly around testing...

ywelsch commented:
@bleskes Pushed another round addressing your comments. Rebased on master so I get #14535. I still cannot remove the line `.put("transport.host", "127.0.0.1")` as there seems to be another issue here. Ports are now correct but binding to multiple interfaces seems to mess with the test.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java:
```diff
@@ -48,6 +50,19 @@
+        public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                             ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+            return newState.nodes().masterNodeId() != null
+                    && !newState.nodes().masterNodeId().equals(previousState.nodes().masterNodeId());
```
can we change this to support the case where master election started and completed between the sampling of the cluster state in the observer and when this function gets called? I think `newState.nodes().masterNodeId() != null
                    && newState != previousState);` is good? (plus a comment)

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterDisruptionIT.java:
```diff
@@ -0,0 +1,115 @@
+        return classes;
+    }
+
+    public void testMasterDisruptionDuringIndexingWithMappingChanges() throws Throwable {
```
can we add a short java doc about what the test does?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterDisruptionIT.java:
```diff
@@ -0,0 +1,115 @@
+
+        // interrupt communication between master and other nodes in cluster
+        // wait a little bit until some data mapping changes are queued
+        Thread.sleep(randomInt(500));
```
why is this needed?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterDisruptionIT.java:
```diff
@@ -0,0 +1,115 @@
+        partition.startDisrupting();
+
+        logger.info("--> waiting for 3 seconds");
+        Thread.sleep(3000);
```
why is this needed? ensureStableCluster(3,dataNode); should be enough imho...

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/IndexingMasterDisruptionIT.java:
```diff
@@ -0,0 +1,115 @@
+                .setSettings(Settings.builder().put("index.number_of_shards", 1).put("index.number_of_replicas", 0))
+                .get();
+        ensureGreen("myindex");
+        client().prepareIndex("myindex", "mytype").setSource("field1", "val1").get();
```
why the extra index? can be removed no?

bleskes commented:
@ywelsch did another round. Let's the leave the networking related code (i.e., `.put("transport.host", "127.0.0.1")` in and change them in another followup. As you indicated other tests have similar code so we can follow up and remove it from them as well.

bleskes commented:
LGTM! :dancers:

bleskes commented:
@ywelsch I think we should back port this to 2.2 .

ywelsch commented:
2.2 does not have `FailedToCommitClusterStateException`...

bleskes commented:
True, but it does have the “failed while in queue” logic. And the code is much cleaner.



> On 10 Nov 2015, at 14:56, Yannick Welsch <notifications@github.com> wrote:
> 
> 2.2 does not have FailedToCommitClusterStateException...
> 
> —
> Reply to this email directly or view it on GitHub.
>

ywelsch commented:
Back ported and pushed to 2.x minus FailedToCommitClusterStateException.

bleskes commented:
Thanks Yannick!

> On 10 Nov 2015, at 19:26, Yannick Welsch <notifications@github.com> wrote:
> 
> Back ported and pushed to 2.x minus FailedToCommitClusterStateException.
> 
> —
> Reply to this email directly or view it on GitHub.
>
